kratos/processes/find_intersected_geometrical_objects_process.cpp

```diff
@@ -25,231 +27,457 @@
-    {
-        this->FindIntersectedSkinObjects(mIntersectedObjects);
-    }
+/***********************************************************************************/
+/***********************************************************************************/
 
-    std::vector<PointerVector<GeometricalObject>>& FindIntersectedGeometricalObjectsProcess::GetIntersections()
-    {
-        return mIntersectedObjects;
-    }
+template<class TEntity>
+void FindIntersectedGeometricalObjectsProcess<TEntity>::Initialize()
+{
+    GenerateOctree();
+}
 
-    ModelPart& FindIntersectedGeometricalObjectsProcess::GetModelPart1()
-    {
-        return mrModelPart1;
-    }
+/***********************************************************************************/
+/***********************************************************************************/
 
-    OctreeBinary<OctreeBinaryCell<Internals::DistanceSpatialContainersConfigure>>* FindIntersectedGeometricalObjectsProcess::GetOctreePointer()
-    {
-        return &mOctree;
+template<>
+void FindIntersectedGeometricalObjectsProcess<Element>::FindIntersectedSkinObjects(std::vector<PointerVector<GeometricalObject>>& rResults)
+{
+    const SizeType number_of_elements = mrModelPart1.NumberOfElements();
+    auto& r_elements_array = mrModelPart1.ElementsArray();
+    OtreeCellVectorType leaves;
+
+    IndexType counter = 0;
+    rResults.resize(number_of_elements);
+    for (auto& r_element : r_elements_array) {
+        leaves.clear();
+        mOctree.GetIntersectedLeaves(r_element, leaves);
+        FindIntersectedSkinObjects(*r_element, leaves, rResults[counter]);
+        ++counter;
     }
+}
 
-    void FindIntersectedGeometricalObjectsProcess::Clear()
-    {
-        mIntersectedObjects.clear();
-    }
+/***********************************************************************************/
+/***********************************************************************************/
 
-    void FindIntersectedGeometricalObjectsProcess::Execute()
-    {
-        GenerateOctree();
-
-        std::vector<OctreeType::cell_type*> leaves;
-        const int number_of_elements = mrModelPart1.NumberOfElements();
-
-        #pragma omp parallel for private(leaves)
-        for (int i = 0; i < number_of_elements; i++)
-        {
-            auto p_element_1 = mrModelPart1.ElementsBegin() + i;
-            leaves.clear();
-            mOctree.GetIntersectedLeaves(*(p_element_1.base()), leaves);
-            MarkIfIntersected(**(p_element_1.base()), leaves);
-        }
+template<>
+void FindIntersectedGeometricalObjectsProcess<Condition>::FindIntersectedSkinObjects(std::vector<PointerVector<GeometricalObject>>& rResults)
+{
+    const SizeType number_of_conditions = mrModelPart1.NumberOfConditions();
+    auto& r_conditions_array = mrModelPart1.ConditionsArray();
+    OtreeCellVectorType leaves;
+
+    IndexType counter = 0;
+    rResults.resize(number_of_conditions);
+    for (auto& r_condition : r_conditions_array) {
```

Comment:
For the sake of avoiding code duplication, I'd create a method `GetIntersectingEntities` which is the one specialized for `Element`and `Condition`. Thus,  the loop becomes something like `for (auto &r_entity : this->GetIntersectingEntities())`. In this way it is not needed to specialize the `FindIntersectedSkinObjects` method.